### ISSUE TITLE
Reference Signal in Axis Class

### ISSUE THREAD

**Issue by CSSFrancis** [hyperspy/hyperspy#2944]
#### Describe the functionality you would like to see.
I think that a lot of good work was done in #2830 but I think that we could maybe think about taking things one step further.

Right now there are still a couple of things we could do to wrap numpy a little closer. 

For example:
```
test_signal = hs.signals.Signal2D(np.ones((3,3,5,6)))
test_signal.axes_manager[0].size=10
```
This bit of code should probably not be possible as the size of the axis should be set by the underlying array. 

To extend this idea a little more, something like adding an axis to a dataset beyond the scope of the underlying data array should not be possible.

```
test_signal = hs.signals.Signal2D(np.ones((3,3,5,6)))
test_signal.axes_manager._append_axis()
```

The solution to this is to make the `Axis` and `AxisManager` classes aware of the associated signal.  In this case we could add the following function to the `BaseDataAxis` class. 

```
@property
def size():
    if self.signal is None:
        return t.Undefined
    else:
        self.signal.data.shape[self.index_in array]
````

Similarly the ragged property for a signal would be defined by:

```
@property
def ragged():
    return self.signal.data.dtype == object:
```

Expanding on the ragged case, the following:
```
test_signal = hs.signals.BaseSignal(np.empty((3,2), dtype=object))
print(test_signal)  # <BaseSignal, title: , dimensions: (|2, 3)>
```
Doesn't particularly make sense as you have both the ragged dimension and non-ragged dimension as a signal axis. If you set `test_signal.ragged= True` the signal changes to `<BaseSignal, title: , dimensions: (2, 3|ragged)>` which is more what I would expect.  I'm having a hard time seeing an instance where the first case would be worth supporting and should be possible. 

If the axes_manager would just call the associated signal and check to see if the signal is ragged it think it would clarify things a little and shift some of the onus to `numpy` for dealing with ragged signals.

Both the axis and axes manager classes could exist without a associated signal but the size, and ragged properties would not None or undefined. 

The ultimate motivation for this is to support #2876.  Better defining these cases would make things like defining a vector axis a little smoother.  Ultimately many of the functions above will have to be broken in the vector case but I think if it is more strongly typed then it might be easier to accomplish this. 

@magnunor @ericpre I would love to hear your thoughts on this.

**Comment by francisco-dlp** [comment]
The problem is that ``Axis`` and ``AxesManager`` can be useful without an associated ``Signal``. In those cases, one would like to set attributes such as ``size``. An alternative could be:

```python
@property
def size():
    if self.signal is None:
        return self._size
    else:
        self.signal.data.shape[self.index_in array]

@size.setter
def size(self, value):
    if self.signal is None:
        self._size = value
    else:
         raise AttributeError("Can't set attribute when the axis is associated to a signal")

```

Regarding ``_append_axis()``, if I  get it right, what you are saying is that it doesn't make sense having that method as the alternative, recreating the ``AxesManager`` with the new axis is more robust. I agree—very often less is more and ``AxesManager`` certainly would benefit from some streamlining. The obvious advantage of removing ``_append_axis`` is that we decrease the chances of bugs due to mismatching array sizes and AxesManager axes.

Regarding ``ragged``, I agree that the current implementation is redundant and too verbose. A ``ragged`` dtype should make it automatically ragged.

**Comment by CSSFrancis** [comment]
Sounds, good.  I will start working on a PR with these changes.

**Comment by ericpre** [comment]
> Similarly the ragged property for a signal would be defined by:
> 
> ```
> @property
> def ragged():
>     return self.signal.data.dtype == object:
> ```

A signal with a numpy array with object dtype is not equivalent to a numpy ragged array

> Expanding on the ragged case, the following:
> 
> ```
> test_signal = hs.signals.BaseSignal(np.empty((3,2), dtype=object))
> print(test_signal)  # <BaseSignal, title: , dimensions: (|2, 3)>
> ```
> 
> Doesn't particularly make sense as you have both the ragged dimension and non-ragged dimension as a signal axis. If you set `test_signal.ragged= True` the signal changes to `<BaseSignal, title: , dimensions: (2, 3|ragged)>` which is more what I would expect. I'm having a hard time seeing an instance where the first case would be worth supporting and should be possible.
> 
> If the axes_manager would just call the associated signal and check to see if the signal is ragged it think it would clarify things a little and shift some of the onus to `numpy` for dealing with ragged signals.

numpy will cast an ragged array to non-ragged when it can. If we rely on this, it means that for the same operation on various dataset, the output can be ragged, but not always, in which case, it is useful to be able to enforce declaring signal as ragged, as it simplify the code.

See #2842 for previous discussion on this.

**Comment by CSSFrancis** [comment]
> numpy will cast an ragged array to non-ragged when it can. If we rely on this, it means that for the same operation on various dataset, the output can be ragged, but not always, in which case, it is useful to be able to enforce declaring signal as ragged, as it simplify the code.

I agree that casting from ragged to non-ragged is something that hyperspy should handle.  But currently to combat that we create an array with a size of 1 and dtype of `object` in most cases or force the data to be a numpy array with dtype ragged.  I don't know if this is always the case though and there are a couple of instances where this is currently failing (see https://github.com/pyxem/pyxem/issues/802).

I can see the benefit of a `to_ragged` function which changes the underlying data, or an `unwrap_ragged`  function which attempts to unwrap a signal, but I don't think that it is possible to create a ragged signal that does not have an object dtype? Maybe I am wrong?  In any case having a ragged signal that is not an object currently would break the map function as well as some others.                        

The reverse case of a signal with an object dtype but not a ragged signal is more complicated.  Maybe there is an instance where this might come into play if you wanted to have a signal axis and a object dtype.  In that case I am not sure what the appropriate behavior is but I don't think forcing the non ragged dimensions to be in the navigation dimensions is too far out of hand.

I guess a better definition of what we consider ragged data would be a good place to start.  Something that I have been struggling with:

- Does a ragged signal have to have an `object` dtype?
    - If not what instances should these be and how do we handle the underlying `numpy` difference 
- If the data array is made up of custom classes (or functions) does that constitute a ragged array
-  Can a ragged array have any signal dimensions?
     - Are (or should) signal dimensions allowed for an array with an `object` dtype (ragged or not)
- Can a ragged array have a single navigation dimension?

**Comment by CSSFrancis** [comment]
@ericpre I've been thinking about this a little more and I wanted to see if you had any more thoughts? I agree with the previous discussion about signals being ragged/non-ragged.  I have no intentions of changing the 0-Dimension behavior but still would like to consider some changes to the ragged code:

Primarily:
- Ragged signals must always have an `object` dtype and signal dimensions=0.  Currently this is (kind of) the case as it is checked  here: https://github.com/hyperspy/hyperspy/blob/5f15a5497d27718adda17a7e89b79fff88207e69/hyperspy/signal.py#L2485-L2487  That being said the following code can create a non-ragged signal that has a ragged dtype. This is me just trying to break things though. 
    - In this case the solution would be to  make the data setter function a little better.  (possibly throwing an error if the number of axes are being changed, or the size of the axes are being changed.  Then again this is most likely an antipattern in which case it might be worth considering my comments above about having the size and the number of axes set by the signal.

```python
x = np.empty(2,dtype=object)
x[0]=np.ones(3)
x[1]=np.ones(4)
sig = hs.signals.BaseSignal(x, ragged=True)
sig.data = np.ones(2)
print(sig.ragged) # True
print(sig.data.dtype) #float64

sig.axes_manager.append_axis() # adding a signal axis to a ragged signal..
print(sig)  # <BaseSignal, title: , dimensions: (2|1ragged)>
```

- Specifically `ragged` signals..  These would be signals which enforce the `object` datatype as well as zero signal dimensions and when created specifically check to see if a datasets has an object datatype.  In the instance that the datatype isn't `object` then it converts the  data associated with the signal axes to an object. This would help the in the case https://github.com/pyxem/pyxem/issues/802 where numpy is casting to a non-object array and it isn't being caught. This might look something like this:

```python
def to_ragged(self):
    if self.data.dtype is "object":
        return
    else:
        new_data = np.empty(self.axes_manager.navigation_shape)
        for i in np.ndindex(self.axes_manager.navigation_shape):
            new_data[i]= self.data[i]
````

**Comment by ericpre** [comment]
Sorry for replying only now, I couldn't do it earlier.

> * Does a ragged signal have to have an `object` dtype?
>   
>   * If not what instances should these be and how do we handle the underlying `numpy` difference

Good question, I am not sure if this is a strict requirement but I imagine that enforcing it would make sense to simplify logic and code.

> * If the data array is made up of custom classes (or functions) does that constitute a ragged array

No, but:
- in numpy, ragged arrays have to have `object` type
- numpy array containing functions have to have `object` type and are not necessary ragged (but they can be ragged too!)

> * Can a ragged array have any signal dimensions?
>   
>   * Are (or should) signal dimensions allowed for an array with an `object` dtype (ragged or not)

In the current implementation, the signal dimension of ragged signal is undefined. Currently a ragged signal is defined as a signal which can contain a variable number of items for different navigation indices. It means that `isig` is no supported in this case. Maybe it is possible to extent it to support signal dimension, I don't know...
In principle, there is no reason that a *non ragged* array with `object` type couldn't be a `Signal2D` or `Signal1D`.
A ragged 

> * Can a ragged array have a single navigation dimension?

Yes, in principle, a ragged signal could have any navigation dimension and this is a not the case, this is a bug!

Regarding your [last comment](https://github.com/hyperspy/hyperspy/issues/2944#issuecomment-1137897019), indeed, this is not good and needs improvement and as you suggested, this could done in the data setter.
What would be the use case for the `to_ragged` method?

**Comment by CSSFrancis** [comment]
No problem with the delay!  

I think from this discussion we have kind of circled in on a couple of things that aren't necessarily changes but are things that we could be doing better with regard to ragged signals.

**Definition of a Ragged Signal**
- Ragged signals must have a `object` dtype.
    - Ragged signals are checked every time the data is changed so that they remain ragged
    - Explicit calling of `signal.ragged=False` must occur for hyperspy to attempt to cast from the `object` dtype
- Ragged signals must have 0 signal dimensions.
    - The function `isig` cannot be called for a ragged signal
    - When setting `signal.ragged=True` all signal axes are transferred to navigation_axes 
- Not all signals with `dtype=object` are ragged arrays.
    - In this case you can have signal axes and hyperspy will return the object at every index in this case.
 
> What would be the use case for the `to_ragged` method?

I was originally thinking that sometimes it might be useful to force a signal to be ragged. In the case of numpy trying to cast something back to a non ragged array a function to "cast-back" might be useful. That might not actually be the best way of doing this and manually setting the `ragged` parameter when building a signal or running a `map` function might be a better idea. 

What I am trying to avoid is the current case in `pyxem` where 0-Dim ragged arrays and N-Dim ragged arrays are treated differently.

**Comment by ericpre** [comment]
Yes, the description of ragged array sounds good to me, maybe it be good to add it to the user guide at some point.

> I was originally thinking that sometimes it might be useful to force a signal to be ragged. In the case of numpy trying to cast something back to a non ragged array a function to "cast-back" might be useful. That might not actually be the best way of doing this and manually setting the `ragged` parameter when building a signal or running a `map` function might be a better idea.

Ok, it sounds like the `to_ragged` method is the same as the `ragged` setter, in which case, we don't need to add a new method, but change the `ragged` setter if necessary.

> What I am trying to avoid is the current case in `pyxem` where 0-Dim ragged arrays and N-Dim ragged arrays are treated differently.

Yes, definitely, this was the motivation for introducing the ragged attribute and also to keep the signal as ragged by default, even in the few case it can not ragged, as it can changed to non-ragged explicitly. It is a bit more verbose, but it has the advantage of avoiding surprised/confusion and simplifying code.
